services/libssh/cve_2018_10933_server.py

```diff
@@ -0,0 +1,333 @@
+            except socket.error as e:
+                if type(e.args) is tuple:
+                    if e.args:
+                        emsg = "{} ({:d})".format(e.args[1], e.args[0])
+                    else:  # empty tuple, e.g. socket.timeout
+                        emsg = str(e) or repr(e)
+                else:
+                    emsg = e.args
+                self._log(ERROR, "Socket exception: " + emsg)
+                self.saved_exception = e
+            except Exception as e:
+                self._log(ERROR, "Unknown exception: " + str(e))
+                self._log(ERROR, util.tb_strings())
+                self.saved_exception = e
+            _active_threads.remove(self)
+            for chan in list(self._channels.values()):
+                chan._unlink()
+            if self.active:
+                self.active = False
+                self.packetizer.close()
+                if self.completion_event is not None:
+                    self.completion_event.set()
+                if self.auth_handler is not None:
+                    self.auth_handler.abort()
+                for event in self.channel_events.values():
+                    event.set()
+                try:
+                    self.lock.acquire()
+                    self.server_accept_cv.notify()
+                finally:
+                    self.lock.release()
+            self.sock.close()
+        except:  # noqa: E722
+            # Don't raise spurious 'NoneType has no attribute X' errors when we
+            # wake up during interpreter shutdown. Or rather -- raise
+            # everything *if* sys.modules (used as a convenient sentinel)
+            # appears to still exist.
+            if self.sys.modules is not None:
+                raise
+
+
+class ParamikoSSHServer(paramiko.ServerInterface):  # noqa: D101
+    def __init__(self):
+        self.event = threading.Event()
+
+    def check_channel_request(self, kind, chanid):  # noqa: D102
+        if kind == "session":
+            return paramiko.OPEN_SUCCEEDED
+        return paramiko.OPEN_FAILED_ADMINISTRATIVELY_PROHIBITED
+
+    def check_auth_password(self, username, password):  # noqa: D102
+        data = {
+            EVENT_TYPE_FIELD_NAME: SSH_ALERT_TYPE,
+            USERNAME_FIELD_NAME: username,
+            PASSWORD_FIELD_NAME: password
+        }
+        self.alert(self.socket, **data)
+        return paramiko.AUTH_FAILED
+
+    def check_auth_publickey(self, username, key):  # noqa: D102
+        data = {
+            EVENT_TYPE_FIELD_NAME: SSH_ALERT_TYPE,
+            USERNAME_FIELD_NAME: username,
+            ADDITIONAL_FIELDS_FIELD_NAME: {
+                KEY_FIELD_NAME: key.get_base64()
+            }
+        }
+        self.alert(self.socket, **data)
+        return paramiko.AUTH_FAILED
+
+    def get_allowed_auths(self, username):  # noqa: D102
+        return "password,publickey"
+
+    def check_channel_shell_request(self, channel):  # noqa: D102
```

Comment:
please provide short description, this is for documentation purposes, excluding D102 is not accepted in this project :)